### PR TITLE
Move to ghactions

### PR DESCRIPTION
**Pull request by tjurtsch**
This submodule implements generic scripts that will be used by CI repos (compilers, misc, prog, eda).

This PR replaced another PR from remote branch  https://github.com/litex-hub/litex-conda-ci/pull/12

### CODE REVIEW

File: README.md
Comment (PiotrZierhoffer) [inline README.md:17]: we need this part as well here

File: action.yml
Comment (PiotrZierhoffer) [inline action.yml:8]: FYI - there's no "allow-failure" mechanism known from Travis. @umarcor are you by any chance aware of something better?

File: after_success.sh
Comment (PiotrZierhoffer) [inline after_success.sh:7]: is there any concept of output folding in gh actions?

File: cleanup-anaconda.sh
Comment (ajelinski) [inline cleanup-anaconda.sh]: `common.sh` containing similar `DATE_SWITCH` variable is sourced; is it necessary to have it also in here? Has it been added for readability?

File: common.sh
Comment (ajelinski) [inline common.sh]: Why such change? Was there any situation where there was an empty `conda-env`?

File: script.sh
Comment (ajelinski) [inline script.sh]: It is installed already in the `conda-env` by `conda-build-prepare` at this point. That environment should be activated by the `common.sh`. I suggest trying without that and using python without it's full path for running `.ci-output.py`.

**Review by ajelinski — COMMENTED** [review]
I suggest dropping `before_cache.sh`, `conda-env.sh` and `conda-meta-extra.sh` files. Currently, they aren't used at all. Please check the `pexpect` thing because it might mean wrong environment is active during building (using python for running `.ci-output.py` might solve the problem).

File: common.sh
Comment (ajelinski) [reply common.sh]: Is it the case of the `master-package.sh` and `cleanup-anaconda.sh`? I think there shouldn't be any `workdir/conda-env` directory at all.

File: wait-for-statuses.py
Comment (PiotrZierhoffer) [inline wait-for-statuses.py:19]: we should verify if it doesn't exceed `max_jobs`

File: wait-for-statuses.py
Comment (PiotrZierhoffer) [inline wait-for-statuses.py:1]: FYI: The aim of this script is to wait for all gh jobs to finish successfully. Travis has "after-success", but there's nothing like that in gh actions. The only way to run a script when all other jobs fail is to

a) mention all jobs in the `needs` node - this is hardly maintainable
b) poll GH actions API for the current status of jobs.

We took the option b), but as the API doesn't list jobs that are not yet started, there's a possibility (encountered in practice) that we poll when all current jobs are finished but there are some yet to be started.

That's why we have a `NUM_OF_JOBS` parameter. We're not very fond of it, but it works for now.

Comment (PiotrZierhoffer) [reply wait-for-statuses.py:1]: again, @umarcor, if you know about some clever tricks here, we're very open to suggestions

File: wait-for-statuses.py
Comment (PiotrZierhoffer) [inline wait-for-statuses.py:32]: we should end the script, failing, whenever there's any failed status

Comment (PiotrZierhoffer) [reply wait-for-statuses.py:32]: or cancelled, or anything that is not about to finish correctly

File: wait-for-statuses.py
Comment (PiotrZierhoffer) [inline wait-for-statuses.py]: I'd go with some more pythonic path joining

File: wait-for-statuses.py
Comment (PiotrZierhoffer) [inline wait-for-statuses.py]:
```suggestion
  print("Completed jobs: " + str(countCompleted) + ". Jobs overall: " + str(numOfJobs))
```

File: cleanup-anaconda.sh
Comment (PiotrZierhoffer) [reply cleanup-anaconda.sh]: I'd drop any repetitions if we have them

**Review by PiotrZierhoffer — CHANGES_REQUESTED** [review]

File: cleanup-anaconda.sh
Comment (tjurtsch) [reply cleanup-anaconda.sh]: It's not necessary. Dropped

File: script.sh
Comment (tjurtsch) [reply script.sh]: I applied running python without it's full path and it works. Removed additional pexpect installation

File: README.md
Comment (tjurtsch) [reply README.md:17]: Readme updated

File: after_success.sh
Comment (tjurtsch) [reply after_success.sh:7]: Yes, there is. In ghactions we can use ::group:: command to fold logs. I applied it where travis_fold was

File: wait-for-statuses.py
Comment (tjurtsch) [reply wait-for-statuses.py:19]: Added suggested check.

File: wait-for-statuses.py
Comment (tjurtsch) [reply wait-for-statuses.py:32]: I've added checks for "non-success" jobs and now we exit with error when any of jobs fails.

File: wait-for-statuses.py
Comment (tjurtsch) [reply wait-for-statuses.py]: Fixed

File: wait-for-statuses.py
Comment (tjurtsch) [reply wait-for-statuses.py]: Fixed

File: common.sh
Comment (tjurtsch) [reply common.sh]: I remember it was needed at some point, but clearly it isn't now. Removed

**Comment by ajelinski** [comment]
LGTM now.

**Review by ajelinski — APPROVED** [review]

**Review by PiotrZierhoffer — APPROVED** [review]

File: action.yml
Comment (umarcor) [reply action.yml:8]: Unfortunately, that's a huge missing feature in GitHub Actions. There is no replacement. You can allow subsequent steps and/or jobs to continue execution after one fails, but the result will be red (failure). AFAIAA, it's not possible to expect failures and have the result green.

In most projects, allowing failures is handled through pytest or some other test runner.

File: after_success.sh
Comment (umarcor) [reply after_success.sh:7]: There is no built-in fold/group timing feature, as in Travis, though. See https://github.com/ghdl/ghdl/blob/master/scripts/ci-run.sh#L65-L75

File: wait-for-statuses.py
Comment (umarcor) [reply wait-for-statuses.py:1]: I'm not really sure about how this is to be used in the wild, but I would blindly argue against a) being hardly maintainable. Sometimes it's better not to fight against the environment, despite it looking poorly featured. Using `needs` in GitHub Actions workflows allows having nice graphs for understanding complex setups. See, for instance https://github.com/ghdl/ghdl/actions/runs/475367314 or https://github.com/verilator/verilator/actions/runs/476444622 or https://github.com/verilog-to-routing/vtr-verilog-to-routing/actions/runs/476502444.

You can handle several dependencies into a single `needs` value by using matrices. So, I would try to look at "jobs" as stages/groups and expand the matrices dynamically if you want to avoid manually maintaining lists of tasks. By doing so, `wait-for-statuses.py` would only be required for checking the statuses of external services (say kokoro).

**Comment by umarcor** [comment]
I'd say I saw some intermediate state of this PR, where the README was "properly" updated. However, some modifications seem gone, and not present after the merge. For example:

> Travis scripts for Litex-Hub/litex-conda-(compilers|eda|misc|prog).

was

> Scripts for Litex-Hub/litex-conda-(compilers|eda|misc|prog) Continuous Integration.

in https://github.com/litex-hub/litex-conda-ci/pull/13/files/ce25d236f590d242de5cb62ae9b1f26c5613f40e#diff-63aa9fac641138cdd078a654078ea278bdf7b39c88d4fc0843304351d6bb3272.

---

Overall, the addition of `action.yml` without further clarification in the README feels very weird. `action.yml` defines a "Composite Action" to be used in GitHub Actions Workflows through `uses: litex-hub/litex-conda-ci@master`. On the other hand, the README says that this repo is to be added as a submodule. Moreover, it seems that the composite action does depend on the submodule being added in a very specific location: https://github.com/litex-hub/litex-conda-ci/blob/master/action.yml#L15-L17. Therefore, I would recommend to convert `action.yml` into `build.sh`. There is no added value by using the YAML syntax in this case. It's an unnecessary layer of complexity which introduces a maintenance burden.

---

I would strongly discourage adding Git for Windows' internal tools to the PATH: https://github.com/litex-hub/litex-conda-ci/blob/master/action.yml#L13. See actions/virtual-environments#1525. What is the motivation for doing so?

File: wait-for-statuses.py
Comment (tjurtsch) [reply wait-for-statuses.py:1]: Thank you for your insight. In Travis CI we used that "stages" approach, where jobs were grouped together. In GitHub Actions we wanted to have these precise dependency relations between jobs. I guess we'll stick with list of tasks for now.

**Comment by tjurtsch** [comment]
Readme clearly needs to be fixed, thank you for pointing that out.
It's also true that this submodule/actions mixture needs to be cleaned up. @ajelinski suggested that we shall drop usage of a submodule and move to full "action" approach instead. It's still not better than a shell script, but it will be cleaner than the current mixed approach. So the litex-conda-ci repo will be only referenced in the workflow as an action. IMHO the only added value will be: no submodule bumping.

**Comment by umarcor** [comment]
@tjurtsch I'm not sure that's the best strategy... Let me explain:

Since GitHub Actions were created, there have been two types: "JavaScript/TypeScript" and "Docker".

JavaScript/TypeScript Actions can be executed on any environment (Ubuntu, Windows or macOS), because all environments include nodejs and someone decided that was the "best" language for scripting. However, those have one very obvious problem: the JS package ecosystem is a mess that fights against any optimal use of files. Unless you want all users of the Action to retrieve hundreds of dependencies, you need to "compile" them. At the same time, GitHub Actions does not support semver filters when specifying Actions, and it can neither use releases. All "releases" need to be a branch in the repo. As a result, you are forced to setting up some CI for handling the releases of your JS Actions, you cannot just tag and push. Alternatively, you need developers to "compile" locally before pushing (so they need nodejs on their workstations/laptops).

Docker Actions are nicer because you can wrap any scripting language (say bash or python) in a container, and you can avoid using a browser language for CI. Unfortunately, Docker Actions are only supported on Ubuntu (neither Windows nor macOS), which is not acceptable in many projects. Moreover, the docker images are fixed in Docker Actions. You cannot let the user pass the image as an argument.

The interesting feature of Docker Actions is that you can call existing containers using that syntax. See, for instance: https://github.com/tmeissner/formal_hw_verification/blob/master/.github/workflows/Test.yml#L41-L45. However, you can only call containers which existed before the job was started. That is, you cannot build a container image in one step and use it in the next one using this syntax. You need to use the regular `docker run ... image_name ...` instead.

Obviously, users complained a lot about the hard to understand limitations, and they recently came up with Composite Actions (the type you used). Those are nice... except they don't allow to compose anything. You cannot call other Actions (use `uses:`) in a Composite Action. Therefore, it's pretty useless, as you can only wrap scripts with it. The only advantage of this type I could find was, precisely, allowing users to customise the container image used in what was earlier a Docker Action: https://github.com/VUnit/vunit_action/blob/master/action.yml.

As a result, I think you should first evaluate whether you'll be able to describe the logic that you currently use in your scripts, considering the constraints above. If you are already following the submodule approach and that works, maybe you should stick with that. I would suggest the following: rework the existing Composite Action for installing this repo in CI and maybe call it `setup-litex-ci`. There, just copy the resources of this repo to some known location, or to an arbitrary location and set an envvar. At the same time, remove the submodules and use this Action as the first step in CI. However, use the scripts explicitly in all other steps of CI. That will allow you to extend this repo for providing different entrypoints, without a maintenance burden.

**Comment by ajelinski** [comment]
Wow, this is quite an extensive explanation.

TBH realizing how many limitations this Composite Action has was very disappointing for me too. It's hard to understand why other Actions can't be called.

However, in this case I see the biggest value of using such an action in dumping the submodule usage. Bumping those is pretty annoying if you have to do it in four repositories. Using its branches to test any changes also isn't the best experience.

Because `litex-hub-ci` isn't a third-party Action we could use its `master` branch in the workflows. The `LiteX-Hub` maintainers decide when it changes so there's no security risk. Maintainability is easier and when there's a need to test some `litex-hub-ci` branch we can easily add an option to the workflows to start the build manually with an input specifying what revision of `litex-hub-ci` is to be used.

You're right that current situation needs to be fixed, because this Action shouldn't use scripts from the repository it runs in. At least there should be `$GITHUB_ACTION_PATH` instead of those paths for it to make sense. I thought that after doing so the `.github/scripts` submodule could be removed from all `litex-conda-*` repositories but I've just realized we have two entrypoints indeed. One for building a package and the other for moving the packages to `main` label if all builds succeeded.

I must say that your proposal with an action simply installing all scripts definitely makes sense and would also enhance the experience of the `litex-conda-ci` usage. However, calling scripts directly isn't an option because there are simply too many jobs in `litex-conda-*` workflow to add more steps. The only option would be to convert the existing `action.yml` to a script and call it although it also adds at least one step to each job – use such `setup-litex-ci` action and then call the build script.

Maybe there could be two actions in this repository:
* build-action
* master-action

It seems to make sense as both actions share some scripts. The problem of accessing two entrypoints without copying the scripts would be solved.

What do you think about such an idea?

**Comment by umarcor** [comment]
@ajelinski, just to make it clear: you are already doing a nice job and all your ideas are sensible. I've been using GitHub Actions workflows for 18 moths, and I currently maintain JS, Docker and Composite actions. I want to help with this background/know-how I have. However, I help handle CI in several orgs, and am not familiar with all the details of the CI infrastructure in SymbiFlow/litex-hub/antmicro (yet). I might be missing relevant constraints, such as what you said about having too many jobs. Therefore, we need to help each other. Can you please provide a reference of the four repositories you mentioned, so I can have a look and get a better picture? I can guess which repos you mean, but I prefer to be given a specific scope to focus on. Maybe we can rethink some of those steps, in coherence with reshaping this repo.

**Comment by ajelinski** [comment]
Sorry for the delayed answer @umarcor. Everything is clear, I hope your help will make `litex-conda-*` CIs even better.

The four repositories that I use `litex-conda-*` name for are:
* https://github.com/litex-hub/litex-conda-compilers
* https://github.com/litex-hub/litex-conda-eda
* https://github.com/litex-hub/litex-conda-misc
* https://github.com/litex-hub/litex-conda-prog

They all use `litex-conda-ci` scripts in their CIs. There's indeed the inconsistency that the action `litex-hub/litex-conda-ci@master` is called and it uses scripts from the calling repository's submodule. Fortunately it doesn't give any negative effects for the package building right now but we'll definitely need to fix it.

BTW. Using separate jobs in the workflows to build each package instead of using a job with a matrix gives us possibility to properly define intra-repository package dependencies. Unfortunately it results in duplicating a similar job multiple times but it seems inevitable if we want to use `needs` in such a way. GitHub Actions disallowing usage of the yaml anchors also doesn't help.

**Pull request merged**
